File: ts-sdk/client/src/gpa/tickArray.ts
//
// Copyright (c) Cryptic Dot
//
// Modification based on Orca Whirlpools (https://github.com/orca-so/whirlpools),
// originally licensed under the Apache License, Version 2.0, prior to February 26, 2025.
//
// Modifications licensed under FusionAMM SDK Source-Available License v1.0
// See the LICENSE file in the project root for license information.
//

import type { Account, Address, GetProgramAccountsApi, GetProgramAccountsMemcmpFilter, Rpc } from "@solana/kit";
import { getAddressEncoder, getBase58Decoder, getI32Encoder } from "@solana/kit";
import { TickArray, TICK_ARRAY_DISCRIMINATOR, FUSIONAMM_PROGRAM_ADDRESS, getTickArrayDecoder } from "../generated";
import { fetchDecodedProgramAccounts } from "./utils";

export type TickArrayFilter = GetProgramAccountsMemcmpFilter & {
  readonly __kind: unique symbol;
};

export function tickArrayStartTickIndexFilter(startTickIndex: number): TickArrayFilter {
  return {
    memcmp: {
      offset: 8n,
      bytes: getBase58Decoder().decode(getI32Encoder().encode(startTickIndex)),
      encoding: "base58",
    },
  } as TickArrayFilter;
}

export function tickArrayFusionPoolFilter(address: Address): TickArrayFilter {
  return {
    memcmp: {
      offset: 113n * 88n + 12n,
      bytes: getBase58Decoder().decode(getAddressEncoder().encode(address)),
      encoding: "base58",
    },
  } as TickArrayFilter;
}

export async function fetchAllTickArrayWithFilter(
  rpc: Rpc<GetProgramAccountsApi>,
  ...filters: TickArrayFilter[]
): Promise<Account<TickArray>[]> {
  const discriminator = getBase58Decoder().decode(TICK_ARRAY_DISCRIMINATOR);
  const discriminatorFilter: GetProgramAccountsMemcmpFilter = {
    memcmp: {
      offset: 0n,
      bytes: discriminator,
      encoding: "base58",
    },
  };
  return fetchDecodedProgramAccounts(
    rpc,
    FUSIONAMM_PROGRAM_ADDRESS,
    [discriminatorFilter, ...filters],
    getTickArrayDecoder(),
  );
}

File: rust-sdk/client/src/core_types/tick_array.rs
//
// Copyright (c) Cryptic Dot
//
// Modification based on Orca Whirlpools (https://github.com/orca-so/whirlpools),
// originally licensed under the Apache License, Version 2.0, prior to February 26, 2025.
//
// Modifications licensed under FusionAMM SDK Source-Available License v1.0
// See the LICENSE file in the project root for license information.
//

use fusionamm_core::{TickArrayFacade, TickFacade};

use crate::{Tick, TickArray};

impl From<TickArray> for TickArrayFacade {
    fn from(val: TickArray) -> Self {
        TickArrayFacade {
            start_tick_index: val.start_tick_index,
            ticks: val.ticks.map(|tick| tick.into()),
        }
    }
}

impl From<Tick> for TickFacade {
    fn from(val: Tick) -> Self {
        TickFacade {
            liquidity_net: val.liquidity_net,
            liquidity_gross: val.liquidity_gross,
            initialized: val.initialized,
            fee_growth_outside_a: val.fee_growth_outside_a,
            fee_growth_outside_b: val.fee_growth_outside_b,
            age: val.age,
            open_orders_input: val.open_orders_input,
            part_filled_orders_input: val.part_filled_orders_input,
            part_filled_orders_remaining_input: val.part_filled_orders_remaining_input,
            fulfilled_a_to_b_orders_input: val.fulfilled_a_to_b_orders_input,
            fulfilled_b_to_a_orders_input: val.fulfilled_b_to_a_orders_input,
        }
    }
}
